influxql/result.go

```diff
@@ -3,8 +3,74 @@ package influxql
 import (
 	"encoding/json"
 	"errors"
+	"hash/fnv"
+	"sort"
 )
 
+// TagSet is a fundamental concept within the query system. It represents a composite series,
+// composed of multiple individual series that share a set of tag attributes.
+type TagSet struct {
+	Tags       map[string]string
+	Filters    []Expr
+	SeriesKeys []string
+	Key        []byte
+}
+
+// AddFilter adds a series-level filter to the Tagset.
+func (t *TagSet) AddFilter(key string, filter Expr) {
+	t.SeriesKeys = append(t.SeriesKeys, key)
+	t.Filters = append(t.Filters, filter)
+}
+
+// Row represents a single row returned from the execution of a statement.
+type Row struct {
+	Name    string            `json:"name,omitempty"`
+	Tags    map[string]string `json:"tags,omitempty"`
+	Columns []string          `json:"columns,omitempty"`
+	Values  [][]interface{}   `json:"values,omitempty"`
+	Err     error             `json:"err,omitempty"`
+}
+
+// tagsHash returns a hash of tag key/value pairs.
+func (r *Row) tagsHash() uint64 {
+	h := fnv.New64a()
+	keys := r.tagsKeys()
+	for _, k := range keys {
+		h.Write([]byte(k))
+		h.Write([]byte(r.Tags[k]))
+	}
+	return h.Sum64()
+}
+
+// tagKeys returns a sorted list of tag keys.
+func (r *Row) tagsKeys() []string {
+	a := make([]string, len(r.Tags))
+	for k := range r.Tags {
+		a = append(a, k)
```

Comment:
This seems wrong. Did you mean to allocate the slice with a capacity (instead of length)?